expose the current active routes list in the Router API

vlukashov opened the pull request:
- add a `Router.activeRoutes` property
 - add to router events' detail object a reference to the router
 - add tests
 - add a demo into the 'Lifecycle callbacks' section

Closes #146

SomeoneToIgnore commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed, 2 unresolved discussions (waiting on @vlukashov)

---

*[demo/vaadin-router-lifecycle-callbacks-demos.html, line 396 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/201#-LFzvXnDCAdPn4MEaJOj:-LFzvXnDCAdPn4MEaJOk:bracjgx) ([raw file](https://github.com/vaadin/vaadin-router/blob/d348b147d526c9a88bcde3eecea1a1c1d90d0035/demo/vaadin-router-lifecycle-callbacks-demos.html#L396)):*
> ```HTML
>               {path: '/users', breadcrumb: {title: 'Users', href: '/users'}, children: [
>                 {path: '/', component: 'x-user-list'},
>                 {path: '/:user', breadcrumb: {title: 'Kim', href: '/users/kim'}, component: 'x-user-profile'},
> ```

Non-blocking:

Not sure if we should it, but to me those `breadcrumb` objects looked like a part of an api.
Especially confusing was this one with the same name for all parameters.

Maybe it would be better to define those through action or at least put a comment that `breadcrumb` is not a part of the router api.

---

*[src/router.js, line 339 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/201#-LFzue6813BC6mmXXe-r:-LFzue6813BC6mmXXe-s:b-f4x7k7) ([raw file](https://github.com/vaadin/vaadin-router/blob/d348b147d526c9a88bcde3eecea1a1c1d90d0035/src/router.js#L339)):*
> ```JavaScript
>         }
>         this.__previousContext = context;
>         this.activeRoutes = context.chain;
> ```

This looks a bit dangerous, can the user change this array accidentally and spoil our resolutions?
A test on this might be also nice.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/201)*
<!-- Sent from Reviewable.io -->

web-padawan commented:
Review status: all files reviewed, 3 unresolved discussions (waiting on @vlukashov)

---

*[demo/vaadin-router-lifecycle-callbacks-demos.html, line 370 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/201#-LFzwVEP-OUj9loKMlBo:-LFzwVEP-OUj9loKMlBp:bxmofn2) ([raw file](https://github.com/vaadin/vaadin-router/blob/d348b147d526c9a88bcde3eecea1a1c1d90d0035/demo/vaadin-router-lifecycle-callbacks-demos.html#L370)):*
> ```HTML
>       property of a <code>router</code> object is a read-only list of routes
>       that correspond to the last completed navigation. This may be useful for
>       example when creating a breadcrumbs componenent.
> ```

typo: component

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/201)*
<!-- Sent from Reviewable.io -->

web-padawan commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed, 3 unresolved discussions (waiting on @vlukashov)

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/201)*
<!-- Sent from Reviewable.io -->

vlukashov commented:
Review status: 1 of 6 files reviewed, 3 unresolved discussions (waiting on @SomeoneToIgnore, @web-padawan, and @vlukashov)

---

*[demo/vaadin-router-lifecycle-callbacks-demos.html, line 370 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/201#-LFzwVEP-OUj9loKMlBo:-LG-P0bQC7WdRWwrhBkT:b-896fix) ([raw file](https://github.com/vaadin/vaadin-router/blob/d348b147d526c9a88bcde3eecea1a1c1d90d0035/demo/vaadin-router-lifecycle-callbacks-demos.html#L370)):*
<details><summary><i>Previously, web-padawan (Sergey Kulikov) wrote…</i></summary><blockquote>

typo: component
</blockquote></details>

Done.

---

*[demo/vaadin-router-lifecycle-callbacks-demos.html, line 396 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/201#-LFzvXnDCAdPn4MEaJOj:-LG-POmO4P1AQ9LBFLZL:b-61vcsy) ([raw file](https://github.com/vaadin/vaadin-router/blob/d348b147d526c9a88bcde3eecea1a1c1d90d0035/demo/vaadin-router-lifecycle-callbacks-demos.html#L396)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

Non-blocking:

Not sure if we should it, but to me those `breadcrumb` objects looked like a part of an api.
Especially confusing was this one with the same name for all parameters.

Maybe it would be better to define those through action or at least put a comment that `breadcrumb` is not a part of the router api.
</blockquote></details>

Good point! I've changed the property name to look less like a standard API (now it's `xBreadcrumbs`), and added a description paragraph to the demo.

---

*[src/router.js, line 339 at r1](https://reviewable.io/reviews/vaadin/vaadin-router/201#-LFzue6813BC6mmXXe-r:-LG-P21N2j59ch3UfQa4:b-896fix) ([raw file](https://github.com/vaadin/vaadin-router/blob/d348b147d526c9a88bcde3eecea1a1c1d90d0035/src/router.js#L339)):*
<details><summary><i>Previously, SomeoneToIgnore (Kirill Bulatov) wrote…</i></summary><blockquote>

This looks a bit dangerous, can the user change this array accidentally and spoil our resolutions?
A test on this might be also nice.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/201)*
<!-- Sent from Reviewable.io -->

SomeoneToIgnore commented:
Reviewed 5 of 5 files at r2, 4 of 5 files at r3.
Review status: 9 of 10 files reviewed, 1 unresolved discussion (waiting on @SomeoneToIgnore and @web-padawan)

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/201)*
<!-- Sent from Reviewable.io -->

SomeoneToIgnore commented:
Reviewed 1 of 5 files at r3.
Review status: all files reviewed, 1 unresolved discussion (waiting on @web-padawan)

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-router/201)*
<!-- Sent from Reviewable.io -->
